drivers/serial/uart_gecko.c

```diff
@@ -9,18 +9,34 @@
 #include <zephyr/drivers/uart.h>
 #include <zephyr/irq.h>
 #include <em_usart.h>
-#include <em_gpio.h>
 #include <em_cmu.h>
 #include <soc.h>
+#include <zephyr/pm/pm.h>
+#include <zephyr/pm/policy.h>
+
+#ifdef CONFIG_PINCTRL
+#include <zephyr/drivers/pinctrl.h>
+#else
+#include <em_gpio.h>
+#endif /* CONFIG_PINCTRL */
 
 #define USART_PREFIX cmuClock_USART
 #define UART_PREFIX cmuClock_UART
 #define CLOCK_USART(id) _CONCAT(USART_PREFIX, id)
 #define CLOCK_UART(id) _CONCAT(UART_PREFIX, id)
 
+#if DT_NODE_HAS_PROP(id, peripheral_id)
+#define GET_GECKO_USART_CLOCK(id) CLOCK_USART(DT_INST_PROP(id, peripheral_id))
+#define GET_GECKO_UART_CLOCK(id) CLOCK_UART(DT_INST_PROP(id, peripheral_id))
+#else
+/* HAL expectation: default USART for UART driver is usart1 */
+#define GET_GECKO_USART_CLOCK(id) CLOCK_USART(1)
+#define GET_GECKO_UART_CLOCK(id) CLOCK_UART(id)
```

Comment:
Is there a reason why USART clock is hard coded to be the UASRT1 clock?
When I am using both USART0 and USART1 the MCU will crash when initializing USART0.
I am using an EFR32PG22 and the following change resolved this issue for me:
```suggestion
#define GET_GECKO_USART_CLOCK(id) CLOCK_USART(id)
#define GET_GECKO_UART_CLOCK(id) CLOCK_UART(id)
```

Reply:
Thank you for pointing that out. This actually strongly depends on how the USART will be used in the system. For example you can use USART0 as SPI and USART1 as UART. In such situation when instantiating the UART driver it will have `instance_id == 0` because it is the first and only UART in the system and that will cause conflicts with straightforward mapping `CLOCK_USART(id)`. You can of course switch the scheme and use USART1 as SPI but then the problem will be moved to the SPI driver. However, this particular setup is valid for Thunderboard EFR32BG22 platform in one specific configuration.
I will remove the hardcoded mapping and replace it with something more robust.

Reply:
I changed the hardcoded values to mapping based on USART base addresses taken from defines in HAL. This should now work for all SiLabs platforms in Zephyr.